### ISSUE TITLE
Find a way to automatically exclude transitive dependencies from autopresets (target platforms)

### ISSUE THREAD

**Issue by izaera** [liferay/liferay-frontend-projects#566]
### Issue type (mark with `x`)

-   [ ] :thinking: Question
-   [ ] :bug: Bug report
-   [X] :gift: Feature request
-   [ ] :woman_shrugging: Other

### Version (mark with `x`)

-   [X] :two: v2.x
-   [ ] :three: v3.x

### Description

With the target platform feature we are making projects depend on a single dependency that transitively brings the whole portal and the devtools needed for the build (babel, bundler, etc.).

Because we add those as `dependencies`, not `devDependencies` (because otherwise the project wouldn't install them in the `node_modules` folder) the bundler thinks they may be needed at runtime and tries to bundle them, leading to huge build times.

The way to avoid this is to use the [exclude](https://github.com/liferay/liferay-frontend-projects/blob/master/maintenance/projects/js-toolkit/docs/.npmbundlerrc-file-reference.md#exclude) config option. However, because we may use semver expressions for devtools, it is possible that the list of exclusion we write today doesn't filter all the deps brought tomorrow (for example if a new version of babel appears that pulls more deps) so we need a way to dinaymically filter the devtool deps brought by the target platform.

So, the best option seems to use a bundler copy plugin (smilar to [liferay-npm-bundler-plugin-exclude-imports](https://github.com/liferay/liferay-frontend-projects/tree/master/maintenance/projects/js-toolkit/packages/liferay-npm-bundler-plugin-exclude-imports)) that is smart enough to find devtools deps and exclude them from the build (unless the user has explicitly added them as deps to the project's package.json).

**Comment by izaera** [comment]
I will defer this for after the target platform is already working (I will add exclusions by hand for now). This way we don't take a diversion from the original goal.

**Comment by izaera** [comment]
Maybe we can simply ignore all deps coming from the autopreset (target platform) when [compiling the dependency graph](https://github.com/liferay/liferay-frontend-projects/blob/2d4996bb625532930b6a6812cfe997b04d83eeb8/maintenance/projects/js-toolkit/packages/liferay-npm-bundler/src/index.ts#L61) in the bundler. Using a copy plugin seems like a complex task due to the big number of transitive deps the target platform pulls...

**Comment by bryceosterhaus** [comment]
> I will defer this for after the target platform is already working

👍 

One thought here is that we might want to make sure to have some thorough testing so that we can test bundles sizes and build times before/after dependency updates.

**Comment by izaera** [comment]
> One thought here is that we might want to make sure to have some thorough testing so that we can test bundles sizes and build times before/after dependency updates.

Yeah. I guess that should be done by QA, but if they don't provide it/have resources to do it, we can design some. One simple way to test if things are OK is to have pre-built JARs committed, then build them in the test and compare.

**Comment by izaera** [comment]
I was thinking that excluding ALL transitive dependencies from autopresets could be too restrictive, because we may truly want to inject some dependencies from the autopreset to be present in the final JAR. So, the problem is how to differentiate those we want to be there (in the JAR) from those we don't. 

But then, realized that we can use the [include-dependencies](https://github.com/liferay/liferay-frontend-projects/blob/master/maintenance/projects/js-toolkit/docs/.npmbundlerrc-file-reference.md#include-dependencies) setting in the autopreset's configuration file to force any dependency we want to be inside the JAR, so this is already covered :relaxed: .

**Comment by faragos** [comment]
Does this mean stuff like babel won't be included in the final builds anymore? For example these dependecies will be removed? 😃 
![image](https://user-images.githubusercontent.com/10152106/121689294-bab8b500-cac4-11eb-88d2-688c36c9dee1.png)

**Comment by izaera** [comment]
@faragos Yes and no. It will be excluded for autopresets, which are a new type of presets. For the older ones, they will be kept unless they are listed as `devDependencies`.

In your case, you can [exclude](https://github.com/liferay/liferay-frontend-projects/blob/master/maintenance/projects/js-toolkit/docs/.npmbundlerrc-file-reference.md#exclude) those  or move them to `devDependencies`, most probably.

**Comment by faragos** [comment]
@izaera So for example if I add "@babel/code-frame" to the devDependencies, even tho it's a transitive one, it will be excluded? Ye I know the exclude feature, but it is not always easy to be sure which one can I exclude. Feels like a little bit of trial and error or am I doing it wrong? 😄

**Comment by izaera** [comment]
Transitive of which? If it is transitive, adding it to `devDependencies` won't help, because it will appear in the dependencies graph. The bundler simply goes to your `package.json`, grabs everything under `dependencies` and traverses it recursively untill it visits the whole graph.

In any case, why is someone adding `@babel/code-frame` as a dep instead of devDep ? :thinking:

**Comment by faragos** [comment]
@izaera True. I checked my dependencies again and I'm not sure tbh if this is correct but:
I use [styled-components](https://github.com/styled-components/styled-components/blob/main/packages/styled-components/package.json) and they have @babel dependencies and they say they should be bundled in (what makes sense to me because they are used inside the files).🤔
So if I understand it correctly. If a library fucks up it's in my duty to fix it with the ignore?
I know this isn't your problem. But I thought maybe it's interesting for you to see what problems I have from the developer prospective😄

**Comment by izaera** [comment]
> it's in my duty to fix it with the ignore?

Exclude, not ignore (ignore is another option only affecting project files :sweat_smile: )

And yes, you are the only one who can fix it, because the bundler has no way to know what needs to be published and which doesn't. This is because the bundler v2 model is "publish everything by default" so that anybody can consume your modules via the AMD loader. So, if the bundler decided to exclude some packages (based on what :thinking: ?) it could break your build if some other bundle depends on them.

Imagine a deployment where you put react in a shared bundle, and then you have three more portlets that import react from that shared bundle (like in the [showcase](https://github.com/izaera/liferay-js-toolkit-showcase) for example). If the bundler applied any heuristic to bundle the react provider it would exclude react because it is not used in that bundle, thus the rest of the projects would fail when deployed.

In other words: the bundler has not enough information during build time to know what should go in the JAR :-(

Webpack has that information, because it traverses deps navigating .js files, not package.json files, thus it can prune the dependency graph without any risk, because the output it's self contained and not used from outside.

We changed the model to be the opposite ("don't publish anything by default") in bundler 3, but since it has not been already released and until we decide what path to follow, it's not a viable alternative.
